README.md
# Lactate

`npm install -g lactate`

An exceedingly fast static file handler, with a few electives.

### Features

* In-memory caching
* Redis caching
* Robust cache-control setting
* Automatic gzipping
* Automatic minification
* Custom 404 pages
* Custom response headers
* Custom / automatic charset headers
* Custom gzip patterns
* Asset bundling and minification
* Middleware export
* Default error pages
* on(status) listeners
* Colored log output
* Global executable
* Directory indexing
* Express.static API compatibility
* node-static API compatibility

## Using Lactate

### Global executable

If installed globally with `npm install -g lactate`, you will have the `lactate` command at your disposal. This will run lactate static file server in the current working directory, utilizing the `cluster` module for multiple CPU cores. All [options](https://github.com/Weltschmerz/Lactate#options) are available.

```code
$ lactate --help
Usage: lactate [options]

Options:
--root, -r                    Local path                          [default: ""]
--from, -f                    Public path                         [default: ""]
--subdirs, -s                 Serve subdirectories                [default: true]
--hidden, -h                  Serve hidden files                  [default: false]
--error_pages, --error-pages  Serve error pages                   [default: true]
--autoindex, -a               Automatically index directories     [default: true]
--cache, -c                   Store assets in-memory              [default: true]
--redis_cache, --rc           Store assets in-memory using Redis  [default: true]
--watch_files, --watch-files  Watch files for cache update        [default: true]
--max_age, --max-age, -M      Client-side caching max-age         [default: 172800]
--gzip, -g                    Gzip text assets                    [default: true]
--minify, -m                  Minify text assets                  [default: false]
--bundle, -b                  Bundle text assets                  [default: false]
--rebundle, --rb              Rebundle assets if modified         [default: true]
--headers, -H                 Custom response headers             [default: ""]
--debug, -d                   Log HTTP info                       [default: true]
--quiet, -q                   Prevent all log output              [default: false]

```

### Programmatic lactating

Lactate can be used with either plain node, or with Express. With Express, Lactate is a drop-in replacement for `static` middleware, but with far more ability. The examples below use Express 2.x API for simplicity. See the [examples](https://github.com/Weltschmerz/Lactate/tree/master/example) for various examples.

```js
var lactate = require('lactate');
var express = require('express');

var app = express.createServer();
app.use(lactate.static(__dirname + '/files'));
```

## Testing Lactate

If installed locally (without -g flag to npm install):

1. `cd` into `~/node_modules/lactate`
2. `npm install ./` to install devDependencies
3. `make test` to run mocha test

##The varieties of Lactate experience

###Creating a Lactate server

```js
var lactate = require('lactate'); 
var options = {root:'files'}; 
var server = lactate.createServer(options); 

server.listen(8080);
```

###Creating a directory handler

```js
var lactate = require('lactate');
var options = {from:'public/path/to/files'};
var files = lactate.dir('files', options);

var http = require('http');
var server = new http.Server;

server.addListener('request', function(req, res) {
  files.serve(req, res);
});

server.listen(8080);
```

###Using directory middleware

```js
var lactate = require('lactate');
var files = lactate.dir('files', options);

var http = require('http');
var server = new http.Server;

server.addListener('request', files.toMiddleware());
server.listen(8080);
```

###Integrating with Express

```js
var lactate = require('lactate');
var files = lactate.dir('files', options);

var express = require('express');
var app = express();

app.use(files.toMiddleware());
app.listen(8080);
```

###Using Express.static API

```js
var lactate = require('lactate');
var express = require('express');
var app = express();

app.use(lactate.static(__dirname + '/files'));
app.listen(8000);
```

###Using node-static API

```js
var lactate = require('lactate');
var files = new(lactate.Server)('./files');

var http = require('http');
var server = new http.Server;

// You could also pass a .toMiddleware()'d function

server.addListener('request', function(req, res) {
  files.serve(req, res);
});

server.listen(8000);
```

###Serving individual files

```js
var lactate = require('lactate');

var fileOptions = {from:'public/directory'};
var files = lactate.dir('files', fileOptions);

var pageOptions = {};
var pages = lactate.dir('pages');

var http = require('http');
var server = new http.Server;

server.addListener('request', function(req, res) {
  if (req.url === '/') {
    pages.serve('index.html', req, res);
  } else {
    files.serve(req, res);
  };
});
server.listen(8000);
```

###Setting options

```js
var lactate = require('lactate');

var fileOptions = {
  from:'public'
};

var files = lactate.dir('files', fileOptions);

files.set('cache', false);
files.disable('gzip');
files.maxAge('ten days');

var express = require('express');
var app = express():
  
app.get('/public/*', files.toMiddleware());
app.listen(8080);
```

###Bundling assets

```js
var lactate = require('lactate');
var options = {
  from:'files'
};

var files = lactate.dir('files', options);

// Combine and minify all scripts to 'common.js'
files.bundle('js', 'common.js', function(err, data) {
// Handle errors
});

// Combine and minify all CSS to 'common.css'
files.bundleCSS('common.css');

var express = require('express');
var app = express();

app.use(files.toMiddleware());
app.listen(8000);
```

###Using custom 404 pages

```js
var lactate = require('lactate');
var files = lactate.dir('files');

files.notFound('path/to/404/page.html');

files.notFound(function(req, res) {
  res.writeHead(404);
  res.end('Woops, 404');
});

var express = require('express');
var app = express();

app.use(files.toMiddleware());
app.listen(8000);
```

###Using custom response headers

```js
var lactate = require('lactate');
var files = lactate.dir('files');

files.setHeader('x-powered-by', 'Lactate');
files.header('x-timestamp', function(req, res) {
  return new Date().toUTCString();
});

var headers = {};
files.headers(headers);
```

###Status listeners

Lactate extends EventEmitter for emitting status code events.

```js
var files = lactate.dir('files');
files.on('404', function(req) {
  console.log('404', req.url);
});
```

##Options

### Setting options

Boolean options may be set using `enable` and `disable` methods. Other options may be set using `set` method with either key/value or an options object.

**Passing to initialization function**

```js
var lactate = require('lactate').Lactate({
  max_age:'two days'
})
```

**Using `set` method**

```js
lactate.set('hidden', true)
```

**Using enable/disable:**

```js
lactate.disable('gzip');
lactate.enable('minify');
```

**Special options methods**

Lactate has some special methods to reduce visual clutter. For setting client-side expiration use `maxAge`

```js
lactate.maxAge('two days');
```

is equivalent to:

```js
lactate.set('max age', 'two days');
```

For setting `custom headers` you may use `setHeader`

```js
lactate.setHeader('x-powered-by', 'Rodent exercise');
```

You can also use a function with `setHeader` for added variance:

```js
lactate.setHeader('x-id', function(req, res) {
  return Math.random().toString(36).substring(2);
}):
```

For defining `charsets`, you may use `Lactate.define`

```js
lactate.define('js', 'application/javascript');
```

or with an object

```js
lactate.define({
  'html': 'text/html; charset=utf-8',
  'js': 'application/javascript'
});
```

**Underscores or spaces**

Use spaces instead of underscores if you prefer:

```js
lactate.disable('max age');
lactate.enable('watch files');
```

### Options available

+ `root` **string**

Local directory from which to serve files. By default, the current working directory.

+ `from` **string**

Public directory exposed to clients. If set, only requests from /*directory* will complete. Contrast this with the `root` option which is the location of files on the serving machine, not necessarily the requested path.

+ `subdirs` **boolean**

By default subdirectories are served. To disable this, set `subdirs` to false.

+ `hidden` **boolean**

Whether or not to serve hidden files. Default is false.

+ `error pages` **boolean**

Enabled by default. When disabled, Lactate will not serve error pages for 404 resposes, etc..

+ `autoindex` **boolean**

Automatically display directory indexes. Disabled by default.

+ `cache` **boolean** or **object**

Keep files in-memory. Enabled by default. For caching options and more information about caching strategy, see [Caching Options](https://github.com/Weltschmerz/Lactate#caching-options).

+ `gzip` **boolean**

If false, disables automatic gzipping for text files (HTML, JS, CSS). Enabled by default.

+ `minify` **boolean**

If true, will automatically minify JavaScript and CSS using [Abridge](https://github.com/Weltschmerz/Abridge). Disabled by default.

+ `watch files` **boolean**

Determines whether Lactate will watch files to update its cache. If this is disabled, then your file cache will not update automatically as files are modified on the server.

+ `headers` **object** or **function**

Sets custom response headers. If the option value is a function, it is a callback which is give (req, res) arguments. This function should return the header value; it is a mapping function.

+ `max age` **number** or **string**

Pass this function a number (of seconds) or a string and appropriate headers will be set for client-side caching. Lactate comes with expiration defaults, such as 'two days' or '5 years and sixteen days' See [Expire](https://github.com/Weltschmerz/Expire) for details.

```js
lactate.set('max_age', 87500)
//87500 seconds
lactate.set('max_age', 'two days')
//172800 seconds
lactate.set'max_age', 'five weeks and one minute and ten seconds')
//3024070 seconds
lactate.set('max_age', 'one year and 2 months and seven weeks and 16 seconds')
//41050028 seconds
```

+ `not found` **string** or **function**

For custom 404 handling. Functions are supplied the response for 100% custom response handling. Otherwise, if set to a string, this option will be treated as an ordinary file path and abide rules for gzipping / in-memory cache.

+ `gzip patterns` **array**

Set custom gzip RegExp patterns. The patterns are matched with mime types. E.g. `/(\+|\/)xml$/` may be used for matching XML documents for gzipping. You may either use:

```js
lactate.set('gzip patterns', []);
```

Or perhaps more conveniently:

```js
//Accepts RegExp or String
lactate.gzip(/pattern/, ...);
```

+ `charset` **boolean** or **string**

Set custom response charset, or automatically detect charset using `node-mime`.

```js
lactate.enable('charset');
// Content-Type: text/html; charset=UTF-8
```

```js
lactate.set('charset', 'utf-8');
// Content-Type: text/html; charset=UTF-8
```

+ `debug` **boolean**

Colored status / msg / path logging, for debugging purposes.

###Caching options

Pass an object to the `cache` option setting. The following fields are accepted and optional:

* `expire` Seconds expiration for cache keys. Keys expire after they are untouched for x-seconds. Default is `15min`.
* `max keys` Maximum number of keys to keep in memory. Default is `Infinity`.
* `max size` Maximum size in MB to keep in cache. Default is `100mb`.
* `segment` Determines the threshold after which to segment a file for streaming instead of traditional writing. Default is `200kb`.

```js
  var options = {};
  
  options.cache = {
    expire:5,
    max_keys:200,
    max_size:2,
    segment:100
  };

  var files = lactate.static('files', options);

  // Or use a string representation. Same as setting for max-age.
  options.cache.expire = 'fifteen minutes';
  files.set('cache', options.cache);
```


## License

MIT



History.md

##History

**v0.13.7**

* No API modifications
* Minification and bundling will now use version 0.5.1 of Abridge. This version removes Java dependencies and no longer spawns child processes. It's much faster. 

**v0.13.6**

* Lactate now sets charsets by default
* Added a test for minification (Note: would prefer faster minification without use of child process)
* Significant refactor

**v0.13.3**

* Added Lactate.define(extension, mimeType). Accepts an object or key/value for defining custom mime types.
* Added new option `charset`. Lactate.enable('charset') for automatically setting charset using `node-mime`, or define a custom charset using Lactate.set('charset', <charset>)

**v0.13.2**

* No API removals
* Add Lactate.gzip(pattern) for extending opts.gzip_patterns. These patterns are tested against mime types.
* Lactate status event listeners are given a `FileRequest` object
* Remove function binding wherever possible
* Reuse `FileRequest` object for Lactate.emit, remove a bunch of *crud* related to the previous solution


lib/lactate.js
// MIT License
// 
// Copyright 2012 Brandon Wilson
// 
// Permission is hereby granted, free of charge, to any person 
// obtaining a copy of this software and associated documentation 
// files (the "Software"), to deal in the Software without 
// restriction, including without limitation the rights to use, 
// copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software
// is furnished to do so, subject to the following conditions: 
// 
// The above copyright notice and this permission notice shall be 
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
// OTHER DEALINGS IN THE SOFTWARE.


var fs          = require('fs');
var URL         = require('url');
var path        = require('path');
var util        = require('util');
var events      = require('events');
var zlib        = require('zlib');
var Suckle      = require('suckle');
var abridge     = require('abridge');
var fraction    = require('fraction');
var expire      = require('expire');

var Cache       = require('./Cache');
var Directory   = require('./Directory');
var Logger      = require('./Logger');
var Responses   = require('./Response');
var FileRequest = require('./FileRequest');

/**
 * Lactate
 *
 * @constructor Lactate
 * @param {Object} options
 */

function Lactate(options) { 

  Responses.apply(this);

  this.log = Logger.createLogger();

  this.cache = Cache.createCache();

  this.opts = {
      root:              process.cwd()
    , from:              '/'
    , hidden:            false
    , not_found:         false
    , error_pages:       true

    //Directory options
    , subdirs:           true
    , autoindex:         false
    , bundle:            false
    , rebundle:          true

    //Caching options
    , cache:             true
    , redis_cache:       false
    , watch_files:       true
    , max_age:           86400 * 2

    //Response options
    , headers:           {}
    , dynamic_headers:   {}
    , gzip:              true
    , gzip_patterns:     []
    , minify:            false
    , charset:           true

    , debug:             false
  };

  if (options) this.set(options);

};

util.inherits(Lactate, events.EventEmitter);

//Extend Lactate prototype with HTTP status codes
require('./status_codes').apply(Lactate.prototype);

//Extend Lactate prototype with mime types
require('./mime_types').apply(Lactate.prototype);

/**
 * Get file extension
 *
 * @param {String} filePath
 * @return String
 */

Lactate.prototype.extension = function(filePath) {
  return path.extname(filePath);
};

/**
 * Parse a URL's pathname and wrap
 * it in a call to decodeURI. This
 * takes care of URL queries and 
 * encoded URLs
 *
 * @param {String} url
 * @return String
 */

Lactate.prototype.parseURL = function(url) {
  if (url.charCodeAt(0) !== 0x2F) {
    url = '/' + url;
  };

  return decodeURI(URL.parse(url).pathname);
};

/**
 * Lactate only cares about GET and 
 * HEAD methods
 *
 * @param {String} method
 * @return Boolean
 */

Lactate.prototype.isValidMethod = function(method) {
  return method === 'GET' || method === 'HEAD';
};

/**
 * Determines if a file path points
 * to a hidden file (prefixed with
 * a dot)
 *
 * @param {String} filePath
 * @return Boolean
 */

Lactate.prototype.isHidden = function(filePath) {
  var basename = path.basename(filePath);
  return basename.charCodeAt(0) === 0x2E;
};

/**
 * Determines if a mime type is
 * textual, and hence should be
 * gzipped
 *
 * @param {String} contentType
 * @return Boolean
 */

Lactate.prototype.isEncodable = function(contentType) {
  var patterns = this._get('gzip_patterns');

  var patternMatch = patterns.length 
  && patterns.some(function(pattern) {
    return pattern.test(contentType)
  });

  return patternMatch || /^text/.test(contentType);
};

/**
 * Determines if a content type
 * is minifiable (compressible)
 *
 * @param {String} contentType
 * @return Boolean
 */

Lactate.prototype.isCompressible = function(contentType) {
  var re = /^(text|application)\/(css|javascript)/i;
  return re.test(contentType);
};

/**
 * Determines if a request has a
 * fresh version of the file, 
 * indicated by if-modified-since
 * header
 *
 * @param {HTTPRequest} req
 * @param {Number} mtime
 * @return Boolean
 */

Lactate.prototype.isCached = function(req, mtime) {
  var ims = req.headers['if-modified-since'];
  return ims === mtime;
};

/**
 * Main entry into file serving. 
 * It all begins here.
 *
 * @param {String, HTTPRequest} url
 * @param {HTTPRequest, HTTPResponse} req
 * @param {HTTPResponse} res
 * @param {Function} [next]
 */

Lactate.prototype.serve = function(url, req, res, next) {
  var root = this._get('root');
  var from = this._get('from');
  var dir  = from, fp;

  // Discern request path relative
  // to root and public options

  if (typeof url === 'string') {
    // First argument is relative
    // file path, join it to the
    // root path
    url = this.parseURL(url);
    fp = path.join(root, url);
  } else {
    // First argument is request
    // object, discern the path
    // from object and join it
    // to the root path
    var len = from.length;

    res = req;
    req = url;
    url = this.parseURL(req.url);
    dir = path.dirname(url);
    fp = url.substring(len);
    fp = path.join(root, fp);
  };

  var request = new FileRequest(
    fp, req, res, null, url, dir
  );

  var error, hasNext = typeof next === 'function';

  if (hasNext) {
    error = next.bind(this, null);
  } else {
    error = errorHandler.bind(this, request);
    function errorHandler(request, status) {
      this['_' + status](request);
    };
  };

  // Prevent invalid request methods
  if (!this.isValidMethod(req.method)) {
    return error(405);
  };

  // Prevent serving hidden files
  if (!this._get('hidden') && this.isHidden(fp)) {
    return error(403);
  };

  // Prevent invalid directory
  if (url.indexOf(from) !== 0) {
    return error(403);
  };

  // Prevent disabled subdirectories
  if (!this._get('subdirs') && dir !== from) {
    return error(403);
  };

  // Autoindex directories
  if (this._get('autoindex') && !this.extension(fp)) {
    return this.serveIndex(fp, req, res, error)
  };

  request.once('error', error);

  // Serve file
  this.serveFile(request);
};

/**
 * Intermediary between #serve and #complete,
 * retrieves an item from the cache if it
 * exists, otherwise does an fs#stat to get
 * file mtime.
 *
 * @param {FileRequest} request
 */

Lactate.prototype.serveFile = function(request) {
  var self = this;

  this.getCache(request, getCacheCallback);

  function getCacheCallback(err, cached) {
    if (!err && cached) {
      var headers = cached.headers;
      request.cached = cached;
      request.headers = headers;
      request.mtime = headers['Last-Modified'];
      self.complete.call(self, request);
    } else { 
      self.stat(request);
    };
  };
};

/**
 * Stat a file path, emit an error event
 * on the condition that the file does 
 * not exist.
 *
 * @param {FileRequest} request
 * @param {Function} fn
 */

Lactate.prototype.stat = function(request) {
  var self = this;

  fs.lstat(request.fp, statCallback);

  function statCallback(err, stat) {
    if (err || !(stat.isFile() || stat.isSymbolicLink())) {
      request.emit('error', 404);
    } else {
      request.mtime = stat.mtime.toUTCString();
      self.complete.call(self, request);

      // Watch file for updates
      if (self._get('watch_files')) {
        self.watchFile(request.fp);
      };
    };
  };
};

/**
 * Conditionallly watch files for changes.
 * On change, simply remove from the cache
 * and allow the next request to pick it up.
 *
 * @param {String} filePath
 */

Lactate.prototype.watchFile = function(filePath) {
  if (!this._get('cache')) return;

  var self = this;

  fs.watch(filePath, function fileWatcher(ev) {
    if (ev === 'change') {
      self.cache.remove(filePath);
      self.emit('cache change', filePath);
    };
  });
};

/**
 * Complete a file request, check
 * mtimes for conditionally sending
 * not-modified response. Otherwise,
 * if the file is cached, serve it.
 * Lastly, read and serve the file.
 *
 * @param {FileRequest} request
 * @param {Object} cached
 * @param {Number} mtime
 */

Lactate.prototype.complete = function(request) {
  var self = this;
  var req    = request.req;
  var res    = request.res;
  var cached = request.cached;
  var mtime  = request.mtime;
  var fp     = request.fp;
  var status = request.status;
  var method = request.method;

  request.mtime = mtime;

  var clientCached = mtime
    && this._get('max_age')
    && this.isCached(req, mtime);

  if (clientCached) {
    // The client has fresh file, send
    // a 'not modified' response
    this._304(request);
  } else if (cached) {
    // File cached from previous request
    var max_size = this.cache.segment;
    var data = cached.read();
    var headers  = cached.headers;

    this.attachHeaders(headers, request);

    res.writeHead(status, headers);

    if (method === 'HEAD') {
      res.end();
      this.emit(status, request);
      return;
    };

    // Stream the cached file in
    // segments if its length 
    // exceeds segment threshold
    if (data.length < max_size) {
      // Send the complete buffer
      res.end(data);
    } else {
      // Stream the cached buffer
      // in segments
      var stream = fraction.createStream(data);
      stream.on('error', console.error);
      stream.pipe(res);
    };

    if (status === 200) {
      request.message = 'Ok-cached';
    };

    // Emit status code event
    this.emit(status, request);
  } else { 
    // Non-cached file request
    this.buildHeaders(request);

    // Send the file
    this.send(request);
  };
};

/**
 * Build a headers object for 
 * the response
 *
 * @param {FileRequest} request
 * @param {Function} fn
 */

Lactate.prototype.buildHeaders = function(request) {

  var filePath = request.fp;
  var req = request.req;

  // Look up file type
  var mimeType = this.mime.lookup(filePath);

  var charset = this._get('charset');

  // Conditionally set charset
  if (charset) {
    var prefix = '; charset=';
    switch (typeof charset) {
      case 'boolean':
        var charsets = this.mime.charsets;
        if (charset = charsets.lookup(mimeType))
          mimeType += prefix + charset;
      break;
      case 'string':
        mimeType += prefix + charset;
      break;
    };
  };

  // Default headers
  var headers = { 'Content-Type': mimeType };

  var encode = this._get('gzip')
  && this.isEncodable(mimeType);

  if (encode) {
    headers['Content-Encoding'] = 'gzip';
  };

  // Headers for 'success'
  // status codes only
  if (request.status < 300) {
    // Use no-store, no-cache, 
    // for non-cached requests,
    // otherwise set max-age
    var maxAge = this._get('max_age');
    var cacheControl = maxAge
    ? 'public, max-age=' + maxAge
    : 'no-store, no-cache';

    // Always set must-revalidate

    cacheControl += ', must-revalidate';

    headers['Last-Modified'] = request.mtime;
    headers['Cache-Control'] = cacheControl;
  };

  // Extend response headers
  // with `headers` option.
  this.attachHeaders(headers);
  this.attachHeaders(headers, request);

  request.headers = headers;
};

/**
 * Attach headers, static or 
 * request-variant
 *
 * @param {Object} headers
 * @param {FileRequest} request
 */

Lactate.prototype.attachHeaders = function(headers, request) {
  var dynamic = !!request;
  var type = dynamic 
    ? 'dynamic_headers' 
    : 'headers';

  var ext = this._get(type);
  var keys = Object.keys(ext);
  var len = keys.length;
  var attach;

  if (len < 1) return;

  if (dynamic) {
    attach = function(header) { 
      var req = request.req;
      var res = request.res;
      return header(req, res); 
    }
  } else {
    attach = function(header) {
      return header; 
    };
  };

  while (len--) {
    var header = keys[len];
    headers[header] = attach(ext[header]);
  };
};

/**
 * Send the file and cache it
 * for future requests
 *
 * @param {FileRequest} request
 * @param {Object} headers
 */

Lactate.prototype.send = function(request) {
  var self    = this;
  var fp      = request.fp;
  var req     = request.req;
  var res     = request.res;
  var status  = request.status;
  var headers = request.headers;

  this.emit(status, request);

  var mux = new Suckle();

  mux.oncomplete(muxCallback);

  // Stream with callback is
  // supplied data such that
  // in-memory cache can be set
  function muxCallback(data, len) {
    headers['Content-Length'] = len;

    res.writeHead(status, headers);
    res.end(data);

    if (self._get('cache')) {
      self.setCache(request, data);
    };
  };

  // On error, respond with
  // 500 internal error
  var error = function error() {
    request.emit('error', 500);
  };

  // Open file readstream
  var rs = fs.createReadStream(fp);
  rs.once('error', error);

  // Detect content-type for
  // automatic minification
  // of text files, if enabled

  var cType = headers['Content-Type'];

  var encode = !!headers['Content-Encoding'];

  var compress = this._get('minify')
  && this.isCompressible(cType);

  // File is not text, pipe it
  // directly to the response
  if (!encode && !compress) { 
    return rs.pipe(mux); 
  };

  // Conditionally minify
  // textual assets
  rs = compress ? abridge.minify(rs) : rs;

  if (!encode) {
    // Gzip is disabled
    rs.pipe(mux);
  } else {
    // Gzip is enabled
    var gzip = zlib.createGzip();
    gzip.pipe(mux);
    rs.pipe(gzip);
  };

};

/**
 * Set cache
 *
 * @param {String} filePath
 * @param {Object} headers
 * @param {Buffer} data
 */

Lactate.prototype.setCache = function(request, data) {
  if (!this._get('cache')) return;
  var filePath = request.fp;
  var headers = request.headers;
  this.cache.set(filePath, headers, data);
};

/**
 * Get cache
 *
 * @param {String} filePath
 * @param {Function} fn
 */

Lactate.prototype.getCache = function(request, fn) {
  if (!this._get('cache')) return fn();
  var filePath = request.fp;
  this.cache.get(filePath, fn);
};

/**
 * Safe get option, replaces spaces
 * with underscores for enhanced 
 * presentation
 *
 * @param {String} key
 * @return option
 */

Lactate.prototype.get = function(key) {
  return this._get(key.replace(/\s/g, '_'));
};

/**
 * Unsafe get, used internally for 
 * enhanced performance
 *
 * @param {String} key
 * @return option
 */

Lactate.prototype._get = function(key) {
  return this.opts[key];
};

/**
 * Set option
 *
 * @param {String} key
 * @param val
 */

Lactate.prototype.set = function(key, val) {

  if (typeof key === 'object') {
    var keys = Object.keys(key);
    return keys.forEach(function(opt) {
      this.set(opt, key[opt]);
    }, this);
  };

  key = key.replace(/\s/g, '_');

  if (!this.opts.hasOwnProperty(key)) {
    return;
  };

  var valType = typeof val;

  switch (key) {
    case 'root':
      val = path.resolve(val);
      break;
    case 'from':
      if (val[0] !== '/') val = '/' + val;
      break;
    case 'header':
    case 'dynamic_header':
      return this.setHeader(key, val);
      break;
    case 'max_age':
      val = valType === 'string' 
        ? expire.getSeconds(val)
        : val
      break;
    case 'cache':
      if (!val) break;

      var opts = valType  === 'object' ? val: {};
      this.cache = Cache.createCache(opts);

      val = true;
      break;
    case 'redis_cache':
      if (!val) break;

      var opts = valType  === 'object' ? val : {};
      opts.redis = true;
      this.cache = Cache.createCache(opts);

      val = true;
      break;
    case 'debug':
      if (!val) break;
      Logger.createDebugger.call(this);
      break;
  };

  this.opts[key] = val;
};

// Set boolean option to true
Lactate.prototype.enable = function(key) {
  this.set(key, true);
};

// Set boolean option to false
Lactate.prototype.disable = function(key) {
  this.set(key, false);
};

// Setter for 'root'
Lactate.prototype.root = function(val) {
  this.set('root', val);
};

// Setter for 'from'
Lactate.prototype.from = function(val) {
  this.set('from', val);
};

// Special setter for 'max_age'
Lactate.prototype.max_age = 
Lactate.prototype.maxAge = function(val) {
  this.set('max_age', val);
};

// Special setter for 'not_found'
Lactate.prototype.notFound = function(val) {
  this.set('not_found', val);
};

// Special setter for 'headers'
Lactate.prototype.header = 
Lactate.prototype.headers = 
Lactate.prototype.setHeader = function(key, val) {
  // Recurse
  if (typeof key === 'object') {
    var headers = Object.keys(key);
    headers.forEach(function(header) {
      this.setHeader(header, key[header]);
    }, this);
    return;
  };

  var map = {
    'string':'headers',
    'function':'dynamic_headers' 
  };

  var type = map[typeof val];

  if (!!type) this.get(type)[key] = val;
};

// Setter for gzip_patterns
Lactate.prototype.gzip = function() {
  var args = Array.prototype.slice.call(arguments);

  args = args.map(function(pattern) {
    if (typeof pattern === 'string') {
      return new RegExp(pattern);
    } else {
      return pattern;
    };
  });

  var opts = this.opts;
  opts.gzip_patterns = opts.gzip_patterns.concat(args);
};

// Define custom mime types
Lactate.prototype.define = function(extension, mimeType) {
  if (typeof extension === 'object') {
    var types = Object.keys(extension);
    types.forEach(function(type) {
      this.define(type, extension[type]);
    }, this);
  } else {
    extension = extension.replace(/^\./, '');
    var defineObj = {};
    defineObj[mimeType] = [extension];
    this.mime.define(defineObj);
  };
};

module.exports.Lactate = function(options) {
  return new Lactate(options);
};

module.exports.file = function(path, req, res, options) {
  var lactate = new Lactate(options);
  return lactate.serve(path, req, res);
};

module.exports.dir = function(directory, options) {
  if (typeof directory === 'string') {
    options = options || {};
    options.root = directory;
  } else {
    options = directory || {};
    options.root = options.root || process.cwd();
  };

  var lactate = new Lactate(options);
  Directory.apply(lactate);
  return lactate;
};

// Replacement for Express.static
module.exports.static = function(dir, from, options) {
  options = options || {};

  if (typeof dir === 'object') {
    options = dir;
    dir = options.root || process.cwd();
  } else {
    switch(typeof from) {
      case 'string':
        options.from = from;
        break;
      case 'object':
        options = from;
        break;
    };
  };

  var lactate = module.exports.dir(dir, options);
  return lactate.toMiddleware();
};

// Adaptors for node-static API
module.exports.serveFile = module.exports.file;
module.exports.Server    = module.exports.dir;

// Create a server
module.exports.createServer = function(options) {
  var handler = module.exports.static(options);
  var server = require('http').createServer(handler);
  return server;
};
